src/Client/HedgedConnections.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+#if defined(OS_LINUX)
+
+#include <functional>
+#include <queue>
+#include <Client/HedgedConnectionsFactory.h>
+#include <Client/IConnections.h>
+
+namespace DB
+{
+
+/** To receive data from multiple replicas (connections) from one shard asynchronously,
+  * The principe of Hedged Connections is used to reduce tail latency:
+  * (if we don't receive data from replica for a long time, we try to get new replica
+  * and send query to it, without cancelling working with previous replica). This class
+  * supports all functionality that MultipleConnections has.
+  */
+class HedgedConnections : public IConnections
+{
+public:
+    struct ReplicaState
+    {
+        Connection * connection = nullptr;
+        std::unordered_map<int, ConnectionTimeoutDescriptorPtr> active_timeouts;
+    };
+
+    struct ReplicaLocation
+    {
+        size_t offset;
+        size_t index;
+    };
+
+    struct OffsetState
+    {
+        std::vector<ReplicaState> replicas;
+        size_t active_connection_count;
```

Comment:
It may be not clear why active_connection_count is not 1. Could you add a comment?